[Pytorch] Add basic c10::optional tests

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47015 [Pytorch] Specialize guts of c10::optional for 32-bit scalars
* **#47014 [Pytorch] Add basic c10::optional tests**
* #47013 [pytorch] Remove c10::nullopt_t::init

Some tests are better than zero tests.

Differential Revision: [D24558649](https://our.internmc.facebook.com/intern/diff/D24558649/)

swolchok committed:
[Pytorch] Add basic c10::optional tests

Some tests are better than zero tests.

Differential Revision: [D24558649](https://our.internmc.facebook.com/intern/diff/D24558649/)

[ghstack-poisoned]

swolchok committed:
add missing #include <array> on "[Pytorch] Add basic c10::optional tests"


Some tests are better than zero tests.

Differential Revision: [D24558649](https://our.internmc.facebook.com/intern/diff/D24558649/)

[ghstack-poisoned]

smessmer commented on c10/test/util/optional_test.cpp:
```diff
@@ -0,0 +1,78 @@
+
+TYPED_TEST(OptionalTest, Initialized) {
+  const auto val = getSampleValue<TypeParam>();
+  typename TestFixture::optional opt((val));
```
nit: `using optional = typename TestFixture::optional` to avoid those typenames?

smessmer commented on c10/test/util/optional_test.cpp:
```diff
@@ -0,0 +1,78 @@
+  const auto val = getSampleValue<TypeParam>();
+  typename TestFixture::optional opt((val));
+  auto copy(opt), moveFrom1(opt), moveFrom2(opt);
+  typename TestFixture::optional move((moveFrom1));
```
missing `std::move`?

smessmer reviewed (approved):
thanks, this was sorely needed

swolchok committed:
Update on "[Pytorch] Add basic c10::optional tests"


Some tests are better than zero tests.

Differential Revision: [D24558649](https://our.internmc.facebook.com/intern/diff/D24558649/)

[ghstack-poisoned]

swolchok committed:
Update on "[Pytorch] Add basic c10::optional tests"


Some tests are better than zero tests.

Differential Revision: [D24558649](https://our.internmc.facebook.com/intern/diff/D24558649/)

[ghstack-poisoned]
